Fix outbound payments memory leak on buggy router

valentinewallace opened the pull request:
Prior to this patch, if we attempted to send a payment or probe to a buggy route, we would error but continue storing the pending outbound payment forever. Attempts to retry would result in a “duplicate payment” error.

In the case of ChannelManager::send_payment, we would also fail to generate a PaymentFailed event, even if the user manually called abandon_payment.

This bug is unlikely to have ever been hit in the wild as most users use LDK’s router. Discovered in the course of adding a new send_to_route API.

Now, we’ll properly generate events and remove the outbound from storage.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2421,6 +2421,9 @@ where
 	/// See `PendingOutboundPayment` documentation for more info.
 	///
 	/// See `ChannelManager` struct-level documentation for lock order requirements.
+	#[cfg(test)]
+	pub(super) pending_outbound_payments: OutboundPayments,
+	#[cfg(not(test))]
```
Could we test with `list_recent_payments` instead of exposing this?

valentinewallace replied:
Yeah true, I have to expose this in #3440 anyway but may as well use the public API where possible.

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1740,12 +1740,26 @@ impl OutboundPayments {
 		NS::Target: NodeSigner,
 		F: Fn(SendAlongPathArgs) -> Result<(), APIError>,
 	{
+		macro_rules! remove_session_privs {
```
I mean, it DRYs the code a bit. I tried removing it but it looked un-DRY. Do you prefer to remove it?

valentinewallace replied:
Ah sorry I misinterpreted that, fixed.

valentinewallace commented:
> Hmm, I wonder if there is any way to detect these cases in retrospect? I.e., could we prune them in remove_stale_payments so that they are dropped after upgrade and not linger forever in the storage of users that happened to hit the bug?

I did consider this, it didn't seem worth it at the time because it means introducing extra code for a weird/unlikely case 🤔  One option would be to remove long-expired `Retryable`s based on their `PaymentParameters`, which is an optional field that was added in 0.0.114. Block height is another option. Let me know if you think this is worth it.

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1460,10 +1460,23 @@ impl OutboundPayments {
+							_ => true
+						}
+					})
+					.map(|(_, (path, session_priv))| (path, session_priv));
```
Oops, will switch this to use `filter_map`

valentinewallace commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -1460,10 +1460,22 @@ impl OutboundPayments {
+							// While a MonitorUpdateInProgress is an Err(_), the payment is still
+							// considered "in flight" and we shouldn't remove it from the
+							// PendingOutboundPayment set.
+							Ok(_) | Err(APIError::MonitorUpdateInProgress) => None,
```
It looks like unfortunately all tests pass with this diff:
```diff
diff --git a/lightning/src/ln/outbound_payment.rs b/lightning/src/ln/outbound_payment.rs
index 68d4125a3..430fb166c 100644
--- a/lightning/src/ln/outbound_payment.rs
+++ b/lightning/src/ln/outbound_payment.rs
@@ -1454,7 +1454,8 @@ impl OutboundPayments {
                                                        // While a MonitorUpdateInProgress is an Err(_), the payment is still
                                                        // considered "in flight" and we shouldn't remove it from the
                                                        // PendingOutboundPayment set.
-                                                       Ok(_) | Err(APIError::MonitorUpdateInProgress) => None,
+                                                       Ok(_) => None,
+                                                       // Ok(_) | Err(APIError::MonitorUpdateInProgress) => None,
                                                        _ => Some((path, session_priv))
                                                }
                                        });
```
Even in the follow-up #3534 when rebased on this PR. Seems pre-existing though.

valentinewallace commented:
Added a debug_assert and renamed a variable:
```diff
diff --git a/lightning/src/ln/outbound_payment.rs b/lightning/src/ln/outbound_payment.rs
index 98efd7b5d..8de47c4ae 100644
--- a/lightning/src/ln/outbound_payment.rs
+++ b/lightning/src/ln/outbound_payment.rs
@@ -1465,7 +1465,8 @@ impl OutboundPayments {
                                self.find_route_and_send_payment(payment_hash, payment_id, route_params, router, first_hops, inflight_htlcs, entropy_source, node_signer, best_block_height, logger, pending_events, send_payment_along_path);
                        },
                        PaymentSendFailure::PartialFailure { failed_paths_retry: Some(mut retry), results, .. } => {
-                               let filtered = results.iter().zip(route.paths.iter().zip(onion_session_privs.iter()))
+                               debug_assert_eq!(route.paths.len(), onion_session_privs.len());
+                               let failed_paths = results.iter().zip(route.paths.iter().zip(onion_session_privs.iter()))
                                        .filter_map(|(path_res, (path, session_priv))| {
                                                match path_res {
                                                        // While a MonitorUpdateInProgress is an Err(_), the payment is still
@@ -1475,7 +1476,7 @@ impl OutboundPayments {
                                                        _ => Some((path, session_priv))
                                                }
                                        });
-                               self.remove_session_privs(payment_id, filtered);
+                               self.remove_session_privs(payment_id, failed_paths);
                                Self::push_path_failed_evs_and_scids(payment_id, payment_hash, &mut retry, route.paths, results.into_iter(), logger, pending_events);
                                // Some paths were sent, even if we failed to send the full MPP value our recipient may
                                // misbehave and claim the funds, at which point we have to consider the payment sent, so

valentinewallace commented:
Ah wait, need one more debug_assert.
